### ISSUE TITLE
[security] Validation not performed when input json is array

### ISSUE THREAD

**Issue by Sroose** [nestjs/nest#2874]
## Current behavior
When you encapsulate any Json input object as an array of that object, it automagically accepts it and will handle each subobject BUT the input validation will be skipped. This is a security issue.

## Input Code
Example using 
```ts
@UsePipes(ValidationPipe)

@IsNotEmpty()
mandatoryField: string
```

and input
```json
[
  {
    "optionalField": "I'm the only field here"
  }
]
```
## Expected behavior
Either an automatic validation of each subobject OR an error telling the object does not validate according to the DTO.

## Possible Solution
Do not accept any object that does not conform the DTO

## Environment

<pre><code>
Nest version: 6.6.6
</code></pre>

**Comment by kamilmysliwiec** [comment]
See :
https://github.com/nestjs/nest/issues/803
https://github.com/nestjs/nest/issues/335
Example solution:
https://github.com/nestjs/nest/issues/803#issuecomment-499088608

**Comment by Sroose** [comment]
This issue is actually not about how to validate arrays. It is the fact that when you do NOT expect arrays but you get an array in, the validation is skipped and the array elements are still being processed. This should not be the default behaviour as it is a security risk.

**Comment by kamilmysliwiec** [comment]
Please, use search next time: https://github.com/nestjs/nest/issues/1192

**Comment by Sroose** [comment]
You're still referring to issues that handle **how to validate arrays**. This issue is about **handling normal input but receiving invalid input** which happens to be encapsulating any data as an array.

(Image as example any current online production case that expect plain simple DTO's)

Assuming input is a simple DTO:
```ts
export class simplest {

    @IsInt()
    readonly myNumber: number;
}
```

expecting an input JSON:
```json
{
  "myNumber": 333
}
```

bad actor posts JSON:
```json
[{
  "myNumber": "DROP TABLE"
}]
```

ACTUAL BEHAVIOUR: Querybuilder or whatever behind will process "DROP TABLE" as the inputvalidation IsInt is skipped

**Comment by kamilmysliwiec** [comment]
I think that I finally understand what you mean. See: https://github.com/typestack/class-validator#passing-options

If you want to restrict unknown objects, pass this option here:
```typescript
new ValidationPipe({ forbidUnknownValues: true });
```

**Comment by Sroose** [comment]
Thanks for keeping an open mind. While this option is probably a solution to the problem, I really feel like this default behaviour is an exploitable bug.

I checked the first NestJS sample (01-CATS-APP)

```json
{
	"name": "Garfield",
	"age": "A",
	"breed": "orange"
}
```
gives
```json
{
    "statusCode": 400,
    "error": "Bad Request",
    "message": [
        {
            "target": {
                "name": "Garfield",
                "age": "A",
                "breed": "orange"
            },
            "value": "A",
            "property": "age",
            "children": [],
            "constraints": {
                "isInt": "age must be an integer number"
            }
        }
    ]
}
```

but
```json
[{
	"name": "Garfield",
	"age": "A",
	"breed": "orange"
}]
```
WORKS, cat is added with age A

I think it's safe to assume most live applications will face the same issue.

**Comment by kamilmysliwiec** [comment]
We rely on `class-validator` and this option is, unfortunately, disabled by default.

To make developers life easier & apps less error prone, I have added an extra check (expected type === transformed type) to ensure that arrays won't pass validation (since they shouldn't be used in combination with regular DTOs and `ValidationPipe` either way). Update is already published (`@nestjs/common@6.6.7`).

Thanks for the explanation!
